kernel/src/log_segment.rs

```diff
@@ -112,13 +124,235 @@ impl LogSegment {
+pub(crate) struct LogSegmentBuilder<'a> {
+    fs_client: &'a dyn FileSystemClient,
+    table_root: &'a Url,
+    checkpoint: Option<CheckpointMetadata>,
+    start_version: Option<Version>,
+    end_version: Option<Version>,
+    reversed_commit_files: bool,
+    omit_checkpoint_files: bool,
+}
+impl<'a> LogSegmentBuilder<'a> {
+    pub(crate) fn new(fs_client: &'a dyn FileSystemClient, table_root: &'a Url) -> Self {
+        LogSegmentBuilder {
+            fs_client,
+            table_root,
+            checkpoint: None,
+            start_version: None,
+            end_version: None,
+            reversed_commit_files: false,
+            omit_checkpoint_files: false,
+        }
+    }
+    /// Optionally provide a checkpoint hint that results from reading the `last_checkpoint` file.
+    pub(crate) fn with_checkpoint(mut self, checkpoint: CheckpointMetadata) -> Self {
+        self.checkpoint = Some(checkpoint);
+        self
+    }
+
+    /// Optionally set the start version of the [`LogSegment`]. This ensures that all commit files
+    /// are at this version or above it.
+    #[allow(unused)]
+    pub(crate) fn with_start_version(mut self, version: Version) -> Self {
+        self.start_version = Some(version);
+        self
+    }
+    /// Optionally set the end version (inclusive) of the [`LogSegment`]. This ensures that all commit files
+    /// and checkpoints are at or below the end version.
+    pub(crate) fn with_end_version(mut self, version: Version) -> Self {
+        self.end_version = Some(version);
+        self
+    }
+
+    /// Optionally specify that the [`LogSegment`] will not have any checkpoint files. It will only
+    /// be made up of commit files.
+    #[allow(unused)]
+    pub(crate) fn omit_checkpoint_files(mut self) -> Self {
+        self.omit_checkpoint_files = true;
+        self
+    }
+    /// Optionally specify that the commits in the [`LogSegment`] will be in order. By default, the
+    /// [`LogSegment`] will reverse the order of commit files, with the latest commit coming first.
+    #[allow(unused)]
+    pub(crate) fn with_reversed_commit_files(mut self) -> Self {
+        self.reversed_commit_files = true;
+        self
+    }
+    /// Build the [`LogSegment`]
+    ///
+    /// This fetches checkpoint and commit files using the `fs_client`.
+    pub(crate) fn build(self) -> DeltaResult<LogSegment> {
+        let Self {
+            fs_client,
+            table_root,
+            checkpoint,
+            start_version,
+            end_version,
+            reversed_commit_files,
+            omit_checkpoint_files,
+        } = self;
+        let log_root = table_root.join("_delta_log/").unwrap();
+        let (mut commit_files, mut checkpoint_files) = match (checkpoint, end_version) {
+            (Some(cp), None) => Self::list_log_files_with_checkpoint(&cp, fs_client, &log_root)?,
+            (Some(cp), Some(version)) if cp.version >= version => {
```

Comment:
I think this should be 
```suggestion
            (Some(cp), Some(end_version)) if cp.version <= end_version => {
```

Reply:
(i think this was existing bug)